lab_5_scrapper/scrapper.py

```diff
@@ -120,45 +250,216 @@ def __init__(self, full_url: str, article_id: int, config: Config) -> None:
+                self.article.topics.append(link['title'])
 
     def unify_date_format(self, date_str: str) -> datetime.datetime:
         """
         Unifies date format
         """
-        pass
+        # implementation for zebra-tv.ru website
+        # return datetime.datetime.strptime(date_str, '%Y-%m-%d %H:%M')
+
+        # implementation for nn.ru website
+        return datetime.datetime.strptime(date_str, '%Y-%m-%dT%H:%M:%S')
 
     def parse(self) -> Union[Article, bool, list]:
         """
         Parses each article
         """
-        pass
+        response = make_request(self.full_url, self.config)
+        article_bs = BeautifulSoup(response.content, 'lxml')
+        self._fill_article_with_text(article_bs)
+        self._fill_article_with_meta_information(article_bs)
+        return self.article
 
 
-def prepare_environment(base_path: Union[Path, str]) -> None:
+def prepare_environment(base_path: Union[Path, str], recursive: bool = False) -> None:
     """
     Creates ASSETS_PATH folder if no created and removes existing folder
     """
-    pass
+    if not recursive:
+        if base_path.exists():
+            shutil.rmtree(base_path)
+        base_path.mkdir(parents=True)
+    else:
+        if not base_path.exists():
+            base_path.mkdir(parents=True)
+
+
+class CrawlerRecursive(Crawler):
+    """
+    Recursive crawler implementation
+    """
+    def __init__(self, config: Config):
+        super().__init__(config)
+        self.crawler_data_path = Path(__file__).parent / 'crawler_data.json'
+        self.last_file_index = 0
+        self.last_index = 0
+        self.start_url = config.get_seed_urls()[0]
+        self.urls = [self.start_url]
+        self.article_urls = []
+
+    def _save_crawler_data(self) -> None:
+        """
+        Saves start_url and collected urls
+        from crawler into a json file
+        """
+        crawler_data = {
+            'last_file_index': self.last_file_index,
+            'last_index': self.last_index,
+            'start_url': self.start_url,
+            'urls': self.urls,
+            'article_urls': self.article_urls
+                        }
+        with open(self.crawler_data_path, 'w', encoding='utf-8') as f:
+            json.dump(crawler_data, f, ensure_ascii=True, indent=4, separators=(', ', ': '))
+
+    def load_crawler_data(self) -> None:
+        """
+        Loads start_url and collected urls
+        from a json file into crawler
+        """
+        if self.crawler_data_path.exists():
+            with open(self.crawler_data_path, 'r', encoding='utf-8') as f:
+                crawler_data = json.load(f)
+            self.last_file_index = crawler_data['last_file_index']
+            self.last_index = crawler_data['last_index']
+            self.start_url = crawler_data['start_url']
+            self.urls = crawler_data['urls']
+            self.article_urls = crawler_data['article_urls']
+
+    def update_file_index(self) -> None:
```

Comment:
i use last_file_index to know, what article has last been parsed and file under which number has been last created.
due to that, if we stop crawler during parsing and then start it again, it would continue on parsing and creating files exactly where it stopped. i can't think of a way to implement it in the crawler itself, as it is connected with the parsing process, which is going later after all the articles have been already found